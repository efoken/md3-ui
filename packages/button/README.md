# @md3-ui/button

## Installation

```sh
// with yarn
yarn add @md3-ui/button

// with npm
npm i @md3-ui/button
```

## Import Components

```jsx
import { Button, ButtonBase, IconButton } from "@md3-ui/button"
```
