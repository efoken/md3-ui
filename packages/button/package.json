{
  "name": "@md3-ui/button",
  "version": "0.2.0",
  "main": "dist/md3-ui-button.cjs.js",
  "module": "dist/md3-ui-button.esm.js",
  "types": "dist/md3-ui-button.cjs.d.ts",
  "license": "MIT",
  "homepage": "https://eikefoken.com/md3-ui/",
  "repository": {
    "type": "git",
    "url": "https://github.com/efoken/md3-ui",
    "directory": "packages/button"
  },
  "keywords": [
    "react",
    "react-component",
    "material design",
    "material you"
  ],
  "files": [
    "dist",
    "src"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@md3-ui/hooks": "^0.2.0",
    "@md3-ui/layout": "^0.2.0",
    "@md3-ui/system": "^0.2.0",
    "@md3-ui/utils": "^0.2.0"
  },
  "devDependencies": {
    "@types/react": "^17.0.52",
    "@types/react-native": "^0.68.9",
    "react": "^17.0.2",
    "react-native": "^0.68.5"
  },
  "peerDependencies": {
    "@types/react": "^17.0.45 || ^18.0.0",
    "@types/react-native": "^0.68.9",
    "react": "^17.0.2 || ^18.0.0",
    "react-native": ">=0.67.4"
  }
}
