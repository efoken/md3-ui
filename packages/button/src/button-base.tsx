import { useEventCallback, useForkRef } from "@md3-ui/hooks"
import {
  OwnerStateProps,
  styled,
  StyleSheet,
  SxProps,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { splitProps, __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  Platform,
  Pressable as RNPressable,
  PressableProps as RNPressableProps,
  StyleProp,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"

export type ButtonBaseProps = Omit<
  RNPressableProps,
  "children" | "disabled" | "style"
> & {
  /** @default false */
  centerRipple?: boolean
  children?: React.ReactNode
  /** @default false */
  disabled?: boolean
  /** @default false */
  disableRipple?: boolean
  focusColor?: string
  /** @default 0.12 */
  focusOpacity?: string
  hoverColor?: string
  /** @default 0.08 */
  hoverOpacity?: number
  pressedColor?: string
  /** @default 0.12 */
  pressedOpacity?: number
  rippleColor?: string
  style?: StyleProp<RNViewStyle>
  styles?: {
    root?: RNViewStyle
    container?: RNViewStyle
  }
  sx?: SxProps
}

export type ButtonBaseStyleKey = keyof NonNullable<ButtonBaseProps["styles"]>

const ButtonBaseRoot = styled(RNPressable, {
  name: "ButtonBase",
  slot: "Root",
})<OwnerStateProps<Pick<ButtonBaseProps, "disabled">>>(
  ({ ownerState }) =>
    Platform.OS === "web" && {
      cursor: ownerState.disabled ? "default" : "pointer",
      outlineWidth: 0,
      overflow: "hidden",
      transition: "box-shadow 200ms linear",
      userSelect: "none",
    },
)

const ButtonBaseContainer = styled(RNView, {
  name: "ButtonBase",
  slot: "Container",
})({
  backgroundColor: "rgba(255, 255, 255, 0)",
  overflow: "hidden",
})

const ButtonBaseHover = styled("span", {
  name: "ButtonBase",
  slot: "Hover",
})({
  bottom: 0,
  left: 0,
  position: "absolute",
  right: 0,
  top: 0,
  transition: "opacity 200ms linear",
  zIndex: -1,
})

const ButtonBaseRippleContainer = styled("span", {
  name: "ButtonBase",
  slot: "RippleContainer",
})()

const ButtonBaseRipple = styled("span", {
  name: "ButtonBase",
  slot: "Ripple",
})<
  OwnerStateProps<
    Required<
      Pick<ButtonBaseProps, "hoverOpacity" | "pressedColor" | "pressedOpacity">
    >
  >
>(({ ownerState, theme }) => ({
  animationDuration: "550ms",
  animationKeyframes: {
    "0%": {
      opacity: 1,
      transform: [{ scale: 0 }],
    },
    "100%": {
      opacity: 0,
      transform: [{ scale: 1 }],
    },
  },
  animationTimingFunction: "ease-in",
  backgroundColor: theme.utils.rgba(
    ownerState.pressedColor,
    ownerState.pressedOpacity - ownerState.hoverOpacity,
  ),
  borderRadius: 999,
  pointerEvents: "none",
  position: "absolute",
  zIndex: -1,
}))

export const ButtonBase = React.forwardRef<RNView, ButtonBaseProps>(
  (inProps, ref) => {
    const {
      centerRipple = false,
      children,
      disabled = false,
      disableRipple = false,
      focusColor: focusColorProp,
      focusOpacity = 0.12,
      hoverColor: hoverColorProp,
      hoverOpacity = 0.08,
      href,
      onKeyDown,
      onKeyUp,
      onPress,
      pressedColor: pressedColorProp,
      pressedOpacity = 0.12,
      rippleColor,
      style,
      styles,
      ...props
    } = useThemeProps({ name: "ButtonBase", props: inProps })

    const theme = useTheme()

    const rootRef = React.useRef<RNView>(null)
    const handleRef = useForkRef(rootRef, ref)

    const [ripples, setRipples] = React.useState<
      { id: number; ripple: React.ReactElement }[]
    >([])

    const [{ backgroundColor = null, ...containerStyle } = {}] = splitProps(
      StyleSheet.flatten(style),
      [
        "backgroundColor",
        "borderRadius",
        "bottom",
        "height",
        "left",
        "margin",
        "marginBottom",
        "marginEnd",
        "marginHorizontal",
        "marginLeft",
        "marginRight",
        "marginStart",
        "marginTop",
        "marginVertical",
        "position",
        "right",
        "top",
        "width",
      ],
    )

    const hoverColor = rippleColor ?? hoverColorProp ?? theme.color.background
    const focusColor = rippleColor ?? focusColorProp ?? hoverColor
    const pressedColor =
      rippleColor ?? pressedColorProp ?? theme.color.background

    const underlayColor =
      // When button has no background color, simply use `rippleColor` as is
      backgroundColor == null
        ? theme.utils.rgba(pressedColor, pressedOpacity)
        : // Otherwise we mix the `rippleColor` with the background color
          theme.utils.mix(
            pressedOpacity,
            pressedColor,
            backgroundColor.toString(),
          )

    const removeRipple = (rippleID: number) => () => {
      setRipples((prevRipples) => prevRipples.filter((r) => r.id !== rippleID))
    }

    const appendRipple = React.useCallback(
      (event: TouchEvent | MouseEvent | {} = {}) => {
        const currentRef = rootRef.current as HTMLDivElement | null

        if (disabled || disableRipple || currentRef == null) {
          return
        }

        const { width, height, left, top } = currentRef.getBoundingClientRect()

        // Get the size of the ripple
        let rippleX: number
        let rippleY: number
        let rippleSize: number

        if (
          centerRipple ||
          (!("clientX" in event) && !("touches" in event)) ||
          ("clientX" in event && event?.clientX === 0 && event?.clientY === 0)
        ) {
          rippleX = Math.round(width / 2)
          rippleY = Math.round(height / 2)
        } else {
          const { clientX, clientY } =
            "touches" in event ? event.touches[0] : event
          rippleX = Math.round(clientX - left)
          rippleY = Math.round(clientY - top)
        }

        if (centerRipple) {
          rippleSize = Math.sqrt((2 * width ** 2 + height ** 2) / 3)
        } else {
          const sizeX =
            Math.max(Math.abs(currentRef.clientWidth - rippleX), rippleX) * 2 +
            2
          const sizeY =
            Math.max(Math.abs(currentRef.clientHeight - rippleY), rippleY) * 2 +
            2
          rippleSize = Math.sqrt(sizeX ** 2 + sizeY ** 2)
        }

        const rippleID =
          ripples.length === 0 ? 1 : ripples[ripples.length - 1].id + 1

        const ripple = (
          <ButtonBaseRipple
            key={rippleID}
            ownerState={{
              hoverOpacity,
              pressedColor,
              pressedOpacity,
            }}
            style={{
              height: rippleSize,
              left: -(rippleSize / 2) + rippleX,
              top: -(rippleSize / 2) + rippleY,
              width: rippleSize,
            }}
            onAnimationEnd={removeRipple(rippleID)}
          />
        )

        setRipples((prevRipples) => [...prevRipples, { id: rippleID, ripple }])
      },
      [
        rootRef,
        disabled,
        disableRipple,
        centerRipple,
        ripples,
        pressedColor,
        pressedOpacity,
        hoverOpacity,
      ],
    )

    const isNonNativeButton = () => {
      const buttonEl = rootRef.current as HTMLElement | null
      return (
        buttonEl?.tagName !== "BUTTON" &&
        !(buttonEl?.tagName === "A" && (buttonEl as HTMLAnchorElement).href)
      )
    }

    const handleKeyDown = useEventCallback((event: React.KeyboardEvent) => {
      if (event.key === "Enter" || event.key === " ") {
        appendRipple()
      }

      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === " "
      ) {
        event.preventDefault()
      }

      onKeyDown?.(event)

      // Keyboard accessibility for non interactive elements
      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === "Enter" &&
        !disabled
      ) {
        event.preventDefault()
        onPress?.(event as any)
      }
    })

    const handleKeyUp = useEventCallback((event: React.KeyboardEvent) => {
      onKeyUp?.(event)

      // Keyboard accessibility for non interactive elements
      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === " " &&
        !event.defaultPrevented
      ) {
        onPress?.(event as any)
      }
    })

    React.useEffect(() => {
      const currentRef = rootRef.current as HTMLDivElement | null

      if (Platform.OS === "web" && currentRef != null) {
        currentRef.addEventListener("mousedown", appendRipple)
        currentRef.addEventListener("touchstart", appendRipple)

        return () => {
          currentRef.removeEventListener("mousedown", appendRipple)
          currentRef.removeEventListener("touchend", appendRipple)
        }
      }

      return () => {}
    }, [appendRipple, handleKeyDown, rootRef])

    const ownerState = {
      disabled,
    }

    const button = (
      <ButtonBaseRoot
        ref={handleRef}
        accessibilityRole={href ? undefined : "button"}
        accessibilityState={{ disabled: disabled || undefined }}
        android_ripple={
          disableRipple
            ? {}
            : {
                borderless: true,
                color: pressedColor,
              }
        }
        disabled={disabled}
        href={href}
        ownerState={ownerState}
        pointerEvents={disabled ? "none" : undefined}
        {...(Platform.OS === "web" && {
          style: [style, styles?.root],
        })}
        {...(Platform.OS === "android" && {
          // For Android we need a wrapping View to cut off the ripple effect.
          // Because of this View and when the button should have
          // `position: absolute`, we need to apply it to the wrapper View and
          // override it here to have `position: relative` with 0px inset.
          style: [
            style,
            styles?.root,
            {
              bottom: 0,
              left: 0,
              marginBottom: 0,
              marginEnd: 0,
              marginLeft: 0,
              marginRight: 0,
              marginStart: 0,
              marginTop: 0,
              position: "relative",
              right: 0,
              top: 0,
            },
          ],
        })}
        {...(Platform.OS === "ios" && {
          style: ({ pressed }) => [
            style,
            styles?.root,
            pressed && !disableRipple && { backgroundColor: underlayColor },
          ],
        })}
        onKeyDown={handleKeyDown}
        onKeyUp={handleKeyUp}
        onPress={onPress}
        {...props}
      >
        {({ hovered, focused }) => (
          <>
            {React.Children.only(children)}
            {Platform.OS === "web" && (
              <ButtonBaseHover
                style={{
                  backgroundColor: focused ? focusColor : hoverColor,
                  opacity: focused ? focusOpacity : hovered ? hoverOpacity : 0,
                }}
              />
            )}
            {Platform.OS === "web" && !disableRipple && (
              <ButtonBaseRippleContainer>
                {ripples.map(({ ripple }) => ripple)}
              </ButtonBaseRippleContainer>
            )}
          </>
        )}
      </ButtonBaseRoot>
    )

    return Platform.OS === "android" ? (
      <ButtonBaseContainer style={[containerStyle, styles?.container]}>
        {button}
      </ButtonBaseContainer>
    ) : (
      button
    )
  },
)

if (__DEV__) {
  ButtonBase.displayName = "ButtonBase"
}
