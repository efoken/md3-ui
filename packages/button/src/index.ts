export { ButtonBase } from "./button-base"
export type {
  ButtonBaseProps,
  ButtonBaseStateCallbackType,
  ButtonBaseStyleKey,
} from "./button-base"
export { ElevatedButton } from "./elevated-button"
export type {
  ElevatedButtonProps,
  ElevatedButtonStyleKey,
} from "./elevated-button"
export { ExtendedFab } from "./extended-fab"
export type { ExtendedFabProps, ExtendedFabStyleKey } from "./extended-fab"
export { Fab } from "./fab"
export type { FabProps, FabStyleKey } from "./fab"
export { FilledButton } from "./filled-button"
export type { FilledButtonProps, FilledButtonStyleKey } from "./filled-button"
export * from "./filled-icon-button"
export * from "./icon-button"
export * from "./outlined-button"
export * from "./outlined-icon-button"
export * from "./text-button"
export * from "./tonal-button"
export * from "./tonal-icon-button"
export * from "./use-button-base-state"
