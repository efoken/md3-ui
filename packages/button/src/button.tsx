import { useBoolean } from "@md3-ui/hooks"
import { Text, TextProvider } from "@md3-ui/layout"
import {
  OwnerStateProps,
  styled,
  SxProps,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  TextStyle as RNTextStyle,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"
import { ButtonBase, ButtonBaseProps } from "./button-base"

export interface ButtonProps extends ButtonBaseProps {
  /**
   * The content of the component.
   */
  children?: React.ReactNode
  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled?: boolean
  /**
   * Element placed before the children.
   */
  icon?: React.ReactElement
  /**
   * Override or extend the styles applied to the component.
   */
  styles?: {
    root?: RNViewStyle
    content?: RNTextStyle
    icon?: RNViewStyle
    label?: RNTextStyle
  }
  /**
   * The system prop that allows defining system overrides as well as
   * additional styles.
   */
  sx?: SxProps
  /**
   * The variant to use.
   * @default "elevated"
   */
  variant?: "elevated" | "filled" | "tonal" | "outlined" | "text"
}

export type ButtonStyleKey = keyof NonNullable<ButtonProps["styles"]>

const ButtonRoot = styled(ButtonBase, {
  name: "Button",
  slot: "Root",
})<OwnerStateProps<Pick<ButtonProps, "disabled" | "icon" | "variant">>>(
  ({ theme, ownerState }) => [
    {
      alignItems: "center",
      borderRadius: 20,
      flexDirection: "row",
      height: 40,
      paddingHorizontal: theme.spacing(3),

      ...(ownerState.variant === "elevated" && {
        ...theme.elevation.level1,
        backgroundColor: theme.color.surface,

        ...(ownerState.disabled && {
          ...theme.elevation.level0,
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level2,
        },

        ":focus": {
          ...theme.elevation.level1,
        },
      }),

      ...(ownerState.variant === "filled" && {
        ...theme.elevation.level0,
        backgroundColor: theme.color.primary,

        ...(ownerState.disabled && {
          ...theme.elevation.level0,
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level1,
        },

        ":focus": {
          ...theme.elevation.level0,
        },
      }),

      ...(ownerState.variant === "tonal" && {
        ...theme.elevation.level0,
        backgroundColor: theme.color["secondary-container"],

        ...(ownerState.disabled && {
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level1,
        },

        ":focus": {
          ...theme.elevation.level0,
        },
      }),

      ...(ownerState.variant === "outlined" && {
        ...theme.elevation.level0,
        backgroundColor: "rgba(255, 255, 255, 0)",
        borderColor: theme.color.outline,
        borderWidth: 1,
        paddingHorizontal: theme.spacing(3) - 1,

        ...(ownerState.disabled && {
          borderColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":focus": {
          borderColor: theme.color.primary,
        },
      }),

      ...(ownerState.variant === "text" && {
        ...theme.elevation.level0,
        minWidth: 48,
        paddingEnd: ownerState.icon ? theme.spacing(2) : theme.spacing(1.5),
        paddingStart: theme.spacing(1.5),
      }),
    },
  ],
)

const ButtonContent = styled(TextProvider, {
  name: "Button",
  slot: "Content",
})<OwnerStateProps<Pick<ButtonProps, "disabled" | "variant">>>(
  ({ theme, ownerState }) => ({
    ...(ownerState.variant === "elevated" && {
      color: theme.color.primary,
    }),

    ...(ownerState.variant === "filled" && {
      color: theme.color["on-primary"],
    }),

    ...(ownerState.variant === "tonal" && {
      color: theme.color["on-secondary-container"],
    }),

    ...(ownerState.variant === "outlined" && {
      color: theme.color.primary,
    }),

    ...(ownerState.variant === "text" && {
      color: theme.color.primary,
    }),

    ...(ownerState.disabled && {
      color: theme.utils.rgba(theme.color["on-surface"], 0.38),
    }),
  }),
)

const ButtonIcon = styled(RNView, {
  name: "Button",
  slot: "Icon",
})<OwnerStateProps<Pick<ButtonProps, "variant">>>(({ theme, ownerState }) => ({
  marginEnd: theme.spacing(1),
  marginStart: ownerState.variant === "text" ? 0 : -theme.spacing(1),
}))

const ButtonLabel = styled(Text, {
  name: "Button",
  slot: "Label",
})(({ theme }) => ({
  ...theme.typescale["label-large"],
  textAlign: "center",
}))

export const Button = React.forwardRef<RNView, ButtonProps>((inProps, ref) => {
  const {
    children,
    disabled = false,
    icon,
    style,
    styles,
    variant = "elevated",
    ...props
  } = useThemeProps({ name: "Button", props: inProps })

  const theme = useTheme()

  const [hovered, handleHover] = useBoolean()
  const [focused, handleFocus] = useBoolean()

  const ownerState = {
    disabled,
    focused,
    hovered,
    icon,
    variant,
  }

  return (
    <ButtonRoot
      ref={ref}
      disabled={disabled}
      ownerState={ownerState}
      rippleColor={
        variant === "elevated"
          ? theme.color.primary
          : variant === "filled"
          ? theme.color["on-primary"]
          : variant === "tonal"
          ? theme.color["on-secondary-container"]
          : variant === "outlined"
          ? theme.color.primary
          : variant === "text"
          ? theme.color.primary
          : undefined
      }
      style={[styles?.root, style]}
      onBlur={handleFocus.off}
      onFocus={handleFocus.on}
      onHoverIn={handleHover.on}
      onHoverOut={handleHover.off}
      {...props}
    >
      <ButtonContent ownerState={ownerState} style={styles?.content}>
        {icon && (
          <ButtonIcon ownerState={ownerState} style={styles?.icon}>
            {React.cloneElement(icon, {
              height: 18,
              width: 18,
            })}
          </ButtonIcon>
        )}
        <ButtonLabel style={styles?.label}>{children}</ButtonLabel>
      </ButtonContent>
    </ButtonRoot>
  )
})

if (__DEV__) {
  Button.displayName = "Button"
}
