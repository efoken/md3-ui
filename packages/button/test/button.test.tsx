import { describeConformance, render, screen } from "@md3-ui/test-utils"
import * as React from "react"
import { Button } from "../src/button"

describe("<Button />", () => {
  describeConformance(<Button>Conformance</Button>, () => ({
    inheritComponent: "div",
    refInstanceof: window.HTMLDivElement,
    skip: ["asProp"],
  }))

  it("should render", () => {
    render(<Button>Hello World</Button>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render an elevated button", () => {
    render(<Button variant="elevated">Hello World</Button>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render a filled button", () => {
    render(<Button variant="filled">Hello World</Button>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render a tonal button", () => {
    render(<Button variant="tonal">Hello World</Button>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render an outlined button", () => {
    render(<Button variant="outlined">Hello World</Button>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render a text button", () => {
    render(<Button variant="text">Hello World</Button>)
    const button = screen.getByRole("button")

    expect(button).toMatchSnapshot()
  })

  it("should render a button with icon", () => {
    render(<Button icon={<span>icon</span>}>Hello World</Button>)
    const button = screen.getByRole("button")
    const icon = button.querySelector('[data-class*="Button-icon"]')

    expect(button).toMatchSnapshot()
    expect(icon).not.toBeNull()
  })

  it("should have a ripple by default", () => {
    render(<Button>Hello World</Button>)
    const button = screen.getByRole("button")

    expect(
      button.querySelector('[data-class*="ButtonBase-rippleContainer"]'),
    ).not.toBeNull()
  })

  it("should disable the ripple", () => {
    render(<Button disableRipple>Hello World</Button>)
    const button = screen.getByRole("button")

    expect(
      button.querySelector('[data-class*="ButtonBase-rippleContainer"]'),
    ).toBeNull()
  })

  it("should automatically change the button to an anchor element when href is provided", () => {
    const { container } = render(
      <Button href="https://google.com">Hello</Button>,
    )
    const button = container.firstChild

    expect(button).toHaveProperty("nodeName", "A")
    expect(button).not.toHaveAttribute("role")
    expect(button).not.toHaveAttribute("type")
    expect(button).toHaveAttribute("href", "https://google.com")
  })
})
