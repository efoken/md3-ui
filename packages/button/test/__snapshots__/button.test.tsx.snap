// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-1fz3o5x r-borderRadius-y47klf r-boxShadow-iidqr2 r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-1guathk r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  data-media="md3-17h1dt4 md3-249f"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(103, 80, 164); display: contents;"
  >
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(103, 80, 164);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="background-color: rgb(103, 80, 164); opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;

exports[`<Button /> should render a button with icon 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-1fz3o5x r-borderRadius-y47klf r-boxShadow-iidqr2 r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-1guathk r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  data-media="md3-17h1dt4 md3-249f"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(103, 80, 164); display: contents;"
  >
    <div
      class="css-view-1dbjc4n r-marginLeft-htfu76 r-marginRight-1kb76zh"
      data-class="Button-icon"
    >
      <span
        height="18"
        width="18"
      >
        icon
      </span>
    </div>
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(103, 80, 164);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-backgroundColor-15l9ap5 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;

exports[`<Button /> should render a filled button 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-15l9ap5 r-borderRadius-y47klf r-boxShadow-cnhr7f r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-1guathk r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  data-media="md3-iwmwhq md3-11mp2qk"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(255, 255, 255); display: contents;"
  >
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(255, 255, 255);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="background-color: rgb(255, 255, 255); opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;

exports[`<Button /> should render a text button 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-borderRadius-y47klf r-boxShadow-cnhr7f r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-minWidth-1v6e3re r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-1guathk r-paddingLeft-f727ji r-paddingRight-j2kj52 r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(103, 80, 164); display: contents;"
  >
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(103, 80, 164);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-backgroundColor-15l9ap5 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;

exports[`<Button /> should render a tonal button 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-oc2s68 r-borderRadius-y47klf r-boxShadow-cnhr7f r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-1guathk r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  data-media="md3-iwmwhq md3-11mp2qk"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(29, 25, 43); display: contents;"
  >
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(29, 25, 43);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="background-color: rgb(29, 25, 43); opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;

exports[`<Button /> should render an elevated button 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-1fz3o5x r-borderRadius-y47klf r-boxShadow-iidqr2 r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-1guathk r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  data-media="md3-17h1dt4 md3-249f"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(103, 80, 164); display: contents;"
  >
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(103, 80, 164);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="background-color: rgb(103, 80, 164); opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;

exports[`<Button /> should render an outlined button 1`] = `
<div
  class="css-cursor-18t94o4 css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-1stjixc r-borderColor-1t01w60 r-borderRadius-y47klf r-borderWidth-rs99b7 r-boxShadow-cnhr7f r-cursor-1loqt21 r-flexDirection-18u37iz r-height-eu3ka r-outlineWidth-10paoce r-overflow-1udh08x r-paddingHorizontal-ju726h r-touchAction-1otgn73 r-transition-imt7jj r-userSelect-lrvibr"
  data-class="Button-root ButtonBase-root"
  data-media="md3-18rahmg"
  role="button"
  tabindex="0"
>
  <div
    style="color: rgb(103, 80, 164); display: contents;"
  >
    <div
      class="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-majxgm r-letterSpacing-8ay8q1 r-lineHeight-rjixqe r-textAlign-q4m81j"
      data-class="Button-label Text-root"
      dir="auto"
      style="color: rgb(103, 80, 164);"
    >
      Hello World
    </div>
  </div>
  <span
    class="r-backgroundColor-15l9ap5 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-hddpjs r-zIndex-1wyyakw"
    data-class="ButtonBase-hover"
    style="opacity: 0;"
  />
  <span
    data-class="ButtonBase-rippleContainer"
  />
</div>
`;
